Fixes #5128 removes snappy from the default installation

kushaldas opened the pull request:
## Status

Ready for review

## Description of Changes

Fixes #5128 

Removes snappy and also removes the related apparmor configuration.

## Testing
- [x] `molecule converge -s libvirt-staging-focal`
- [x] `molecule login -s libvirt-staging-focal app-staging`
- [x] check `snapd` package is not present
- [x] repeat the same for mon-staging
- [x] Now repeat the steps for `Xenial` 

## Deployment

Any special considerations for deployment? Consider both:

1. Upgrading existing production instances.
2. New installs.

## Checklist

### If you made changes to the server application code:

- [ ] Linting (`make lint`) and tests (`make test`) pass in the development container

### If you made changes to `securedrop-admin`:

- [ ] Linting and tests (`make -C admin test`) pass in the admin development container

### If you made changes to the system configuration:

- [ ] [Configuration tests](https://docs.securedrop.org/en/latest/development/testing_configuration_tests.html) pass

### If you added or removed a file deployed with the application:

- [ ] I have updated AppArmor rules to include the change

### If you made non-trivial code changes:

- [ ] I have written a test plan and validated it for this PR

Choose one of the following:

- [ ] I have opened a PR in the [docs repo](https://github.com/freedomofpress/securedrop-docs) for these changes, or will do so later
- [ ] I would appreciate help with the documentation
- [ ] These changes do not require documentation

### If you added or updated a code dependency:

Choose one of the following:

- [ ] I have performed a diff review and pasted the contents to [the packaging wiki](https://github.com/freedomofpress/securedrop-debian-packaging/wiki)
- [ ] I would like someone else to do the diff review

kushaldas committed:
Fixes #5128 removes snappy from the default installation

Also removes the related apparmor configuration.

conorsch commented:
Had a chat with @zenmonkeykstop, and I'll pick this one up for review, since I'm updating tests for related PRs already.

conorsch commented on install_files/ansible-base/roles/common/defaults/main.yml:
```diff
@@ -49,3 +49,4 @@ unused_packages:
   - libiw30
   - wireless-tools
   - wpasupplicant
+  - snapd
```
We have config tests for package absence, that should be updated to confirm snapd is not present (on any system).

conorsch commented on install_files/ansible-base/roles/common/tasks/remove_unused_packages.yml:
```diff
@@ -7,6 +7,12 @@
+  file:
+    path: /etc/apparmor.d/usr.lib.snapd.snap-confine.real
+    state: absent
+
```
I just checked, and an `apt-get purge snapd` command removes this file even if it was already on disk. Since the removal above results in a purge, let's write a _test_ for absence of this file to ensure it's doing what we expect.

conorsch reviewed (changes requested):
Works! Requesting more tests to verify the end state, though. I'll tack those on shortly and confirm CI still passing.

rmol commented:
Aren't there snaps installed by default that we should remove before purging snapd? I see at least these on a fresh staging machine:
```
sdadmin@app-staging:~$ snap list
Name    Version   Rev    Tracking       Publisher   Notes
core18  20210128  1988   latest/stable  canonical✓  base
lxd     4.0.5     19188  4.0/stable/…   canonical✓  -
snapd   2.48.2.1  11036  latest/stable  canonical✓  snapd
```

I thought I saw more on a recent hardware installation, but will have to reinstall to double-check.

Commit:
Adds tests for snapd removal

Ensures packages and files are absent on all platforms.

rmol commented:
I was wrong, fresh Focal install on hardware has the same list: `core18`, `lxd`, `snapd`.

conorsch commented:
@rmol My understanding is that once the service is removed, those snaps are dead dir trees. I didn't explicitly check whether a purge removes those snaps, that'd be a great thing to add to the tests. Thanks, will do.

conorsch commented:
Purging does indeed remove the on-disk snaps:

```
sdadmin@app-staging:~$ snap list
Name    Version   Rev    Tracking       Publisher   Notes
core18  20210128  1988   latest/stable  canonical✓  base
lxd     4.0.5     19188  4.0/stable/…   canonical✓  -
snapd   2.48.2.1  11036  latest/stable  canonical✓  snapd
sdadmin@app-staging:~$ ls -l /var/lib/snapd/snaps/
total 320060
-rw------- 2 root root 57626624 Jul 31  2020 core18_1880.snap
-rw------- 1 root root 58159104 Feb 11 21:30 core18_1988.snap
-rw------- 2 root root 74735616 Jul 31  2020 lxd_16099.snap
-rw------- 1 root root 73277440 Feb 11 21:30 lxd_19188.snap
drwxr-xr-x 2 root root     4096 Jul 10  2020 partial
-rw------- 1 root root 32595968 Feb 11 21:30 snapd_11036.snap
-rw------- 2 root root 31334400 Jul 31  2020 snapd_8542.snap
sdadmin@app-staging:~$ sudo apt-get purge snapd -y
Reading package lists... Done
Building dependency tree       
Reading state information... Done
The following package was automatically installed and is no longer required:
  squashfs-tools
Use 'sudo apt autoremove' to remove it.
The following packages will be REMOVED:
  snapd*
0 upgraded, 0 newly installed, 1 to remove and 0 not upgraded.
After this operation, 126 MB disk space will be freed.
(Reading database ... 71227 files and directories currently installed.)
Removing snapd (2.48+20.04) ...
Stopping snap.lxd.activate.service
Stopping unit snap.lxd.activate.service
Waiting until unit snap.lxd.activate.service is stopped [attempt 1]
snap.lxd.activate.service is stopped.
Stopping snap.lxd.daemon.service
Stopping unit snap.lxd.daemon.service
Waiting until unit snap.lxd.daemon.service is stopped [attempt 1]
snap.lxd.daemon.service is stopped.
Stopping snap.lxd.daemon.unix.socket
Stopping unit snap.lxd.daemon.unix.socket
Waiting until unit snap.lxd.daemon.unix.socket is stopped [attempt 1]
snap.lxd.daemon.unix.socket is stopped.
Processing triggers for man-db (2.9.1-1) ...
Processing triggers for dbus (1.12.16-2ubuntu2.1) ...
Processing triggers for mime-support (3.64ubuntu1) ...
(Reading database ... 71146 files and directories currently installed.)
Purging configuration files for snapd (2.48+20.04) ...
Stopping snap.lxd.activate.service
Stopping unit snap.lxd.activate.service
Waiting until unit snap.lxd.activate.service is stopped [attempt 1]
snap.lxd.activate.service is stopped.
Removing snap.lxd.activate.service
Stopping snap.lxd.daemon.service
Stopping unit snap.lxd.daemon.service
Waiting until unit snap.lxd.daemon.service is stopped [attempt 1]
snap.lxd.daemon.service is stopped.
Removing snap.lxd.daemon.service
Stopping snap-core18-1880.mount
Stopping unit snap-core18-1880.mount
Waiting until unit snap-core18-1880.mount is stopped [attempt 1]
snap-core18-1880.mount is stopped.
Removing snap core18 and revision 1880
Removing snap-core18-1880.mount
Stopping snap-core18-1988.mount
Stopping unit snap-core18-1988.mount
Waiting until unit snap-core18-1988.mount is stopped [attempt 1]
snap-core18-1988.mount is stopped.
Removing snap core18 and revision 1988
Removing snap-core18-1988.mount
Stopping snap-lxd-16099.mount
Stopping unit snap-lxd-16099.mount
Waiting until unit snap-lxd-16099.mount is stopped [attempt 1]
snap-lxd-16099.mount is stopped.
Removing snap lxd and revision 16099
Stopping unit snap.lxd.daemon.unix.socket
Waiting until unit snap.lxd.daemon.unix.socket is stopped [attempt 1]
snap.lxd.daemon.unix.socket is stopped.
Stopping unit snap.lxd.daemon.unix.socket
Warning: The unit file, source configuration file or drop-ins of snap.lxd.daemon.unix.socket changed on disk. Run 'systemctl daemon-reload' to reload units.
Waiting until unit snap.lxd.daemon.unix.socket is stopped [attempt 1]
snap.lxd.daemon.unix.socket is stopped.
Removing snap-lxd-16099.mount
Stopping snap-lxd-19188.mount
Stopping unit snap-lxd-19188.mount
Waiting until unit snap-lxd-19188.mount is stopped [attempt 1]
snap-lxd-19188.mount is stopped.
Removing snap lxd and revision 19188
Removing snap-lxd-19188.mount
Stopping snap-snapd-11036.mount
Stopping unit snap-snapd-11036.mount
Waiting until unit snap-snapd-11036.mount is stopped [attempt 1]
snap-snapd-11036.mount is stopped.
Removing snap snapd and revision 11036
Removing snap-snapd-11036.mount
Stopping snap-snapd-8542.mount
Stopping unit snap-snapd-8542.mount
Waiting until unit snap-snapd-8542.mount is stopped [attempt 1]
snap-snapd-8542.mount is stopped.
Removing snap snapd and revision 8542
Removing snap-snapd-8542.mount
Final directory cleanup
Discarding preserved snap namespaces
Removing extra snap-confine apparmor rules
Removing snapd cache
Removing snapd state
sdadmin@app-staging:~$ ls -l /var/lib/snapd/snaps/
ls: cannot access '/var/lib/snapd/snaps/': No such file or directory
```

I've added a test to confirm that path does not exist.

conorsch commented:
Output above from the purge action includes:

> Warning: The unit file, source configuration file or drop-ins of snap.lxd.daemon.unix.socket changed on disk. Run 'systemctl daemon-reload' to reload units.

which explains the CI failure in https://app.circleci.com/pipelines/github/freedomofpress/securedrop/1861/workflows/d67b6601-bcc1-42d4-9dd9-1c53a3ad7ff7/jobs/50275. I've pushed a commit to resolve, will monitor CI to make sure it's passing.

Commit:
Make sure to purge unwanted packages

If we don't purge them, config files can linger, causing problems with
other services. Discovered during removal of snapd, which will leave
apparmor profiles (as well as the old snaps themselves) on disk if only
removed, but not purged.

conorsch commented:
> Purging does indeed remove the on-disk snaps

Ah, I was wrong to assume that we were purging during the removal! We are now, which should resolve the CI failures.

conorsch reviewed (approved):
Approving provisionally, contingent on CI passing.

emkll reviewed (approved):
Went through the test plan locally, and the changes are working as expected. Thanks for adding tests here.
